File: TMB examples/map argument/linear_mixed_model.R

setwd( "C:/Users/James.Thorson/Desktop/Project_git/2016_Spatio-temporal_models/TMB examples/map argument" )

######################
# Simulate data
######################

# Simulate predictors
Factor = rep( 1:10, each=10)
Z = rnorm( length(unique(Factor)), mean=0, sd=1)
X0 = 0

# Simulate response
Y = Z[Factor] + X0 + rnorm( length(Factor), mean=0, sd=1)

######################
# Run in TMB
######################

library(TMB)

# Compile model
Version = "linear_mixed_model"
compile( paste0(Version,".cpp") )

# Build inputs
Data = list( "n_data"=length(Y), "n_factors"=length(unique(Factor)), "Factor"=Factor-1, "Y"=Y)
Parameters = list( "X0"=-10, "log_SD0"=2, "log_SDZ"=2, "Z"=rep(0,Data$n_factor) )

# Turn off random effects
Map = list()
Map[["log_SDZ"]] = factor(NA)
Map[["Z"]] = factor( rep(NA,Data$n_factor) )

# Build object
dyn.load( dynlib("linear_mixed_model") )
Obj = MakeADFun(data=Data, parameters=Parameters, map=Map)  #

# Prove that function and gradient calls work
Obj$fn( Obj$par )
Obj$gr( Obj$par )

# Optimize
start_time = Sys.time()
Opt = nlminb( start=Obj$par, objective=Obj$fn, gradient=Obj$gr, control=list("trace"=1) )

# Get reporting and SEs
Report = Obj$report()
  SD = sdreport( Obj )

File: TMB examples/map argument/linear_mixed_model.cpp
#include <TMB.hpp>
template<class Type>
Type objective_function<Type>::operator() ()
{
  // Data
  DATA_INTEGER( n_data );
  DATA_INTEGER( n_factors );
  DATA_IVECTOR( Factor );
  DATA_VECTOR( Y );                
  
  // Parameters
  PARAMETER( X0 );
  PARAMETER( log_SD0 );
  PARAMETER( log_SDZ );
  PARAMETER_VECTOR( Z );
  
  // Objective funcction
  Type jnll = 0;
  
  // Probability of data conditional on fixed and random effect values
  for( int i=0; i<n_data; i++){
    jnll -= dnorm( Y(i), X0 + Z(Factor(i)), exp(log_SD0), true );
  }
  
  // Probability of random coefficients
  for( int i=0; i<n_factors; i++){
    jnll -= dnorm( Z(i), Type(0.0), exp(log_SDZ), true );
  }
  
  // Reporting
  Type SDZ = exp(log_SDZ);
  Type SD0 = exp(log_SD0);

  REPORT( SDZ );
  REPORT( SD0 );
  REPORT( Z );
  REPORT( X0 );

  ADREPORT( SDZ );
  ADREPORT( SD0 );
  ADREPORT( Z );
  ADREPORT( X0 );

  return jnll;
}
